### PR TITLE
Add SSL Certificate

### PR DESCRIPTION
**Pull request by piadelosreyes**
## Summary of changes

- _Add script to create SSL key and certificate_

/fyi: @tjoelsson, @mark-juwai : Please review.

Thanks

### CODE REVIEW

File: tasks/ssl_certificate.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+---
+- include_vars: "group_vars/env-production/vault.yml"
+
+- name: create admin.juwai.com + digicert certificate file
+  template:
+    src: "admin.juwai.com.crt.j2"
+    dest: "/etc/ssl/admin.juwai.com.crt"
```
Comment (tjoelsson) [inline tasks/ssl_certificate.yml]: Use variable

File: defaults/main.yml [hunk 2]
```diff
@@ -4,3 +4,7 @@ deploy_group: '{{ ansible_ssh_user }}'
 
 laravel_rotate_nginx_log: true
 laravel_rotate_nginx_log_retention: 30
+
+laravel_https_enabled: 'on'
```
Comment (tjoelsson) [inline defaults/main.yml]: Let's make this variable a boolean (yes/no) so adding certificate files and using https is optional. Not all applications will need it, but should still be able to use this role.

File: tasks/ssl_certificate.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    group: root
+    mode: 0600
+
+- name: create admin.juwai.com key file
+  copy:
+    content: "{{ admin_juwai_com_key }}"
+    dest: "/etc/ssl/admin.juwai.com.key"
+    owner: root
+    group: '{{ web_server_user }}'
+    mode: 0640
```
Comment (tjoelsson) [inline tasks/ssl_certificate.yml]: We only want root to able to read this file. 640 is only for a special case where PHP needs to read the file too.

File: templates/admin.juwai.com.crt.j2 [hunk 4]
```diff
@@ -0,0 +1 @@
+{{ admin_juwai_com_crt }}{{ digicert_crt }}
```
Comment (tjoelsson) [inline templates/admin.juwai.com.crt.j2]: Name of this file should be generic

File: tasks/ssl_certificate.yml (hunk 3 above)
Comment (tjoelsson) [reply tasks/ssl_certificate.yml]: Also change the group to root.

File: templates/nginx.conf.j2 [hunk 5]
```diff
@@ -17,6 +17,10 @@ server {
     access_log /var/log/nginx/{{ application_name }}/access.log main;
     error_log /var/log/nginx/{{ application_name }}/error.log;
 
+    ssl                         {{ laravel_https_enabled }};
+    ssl_certificate             {{ laravel_https_certificate }};
+    ssl_certificate_key         {{ laravel_https_key }};
```
Comment (tjoelsson) [inline templates/nginx.conf.j2]: Wrap all these in an if-statement.

**Review by tjoelsson — CHANGES_REQUESTED** [review]

File: defaults/main.yml (hunk 2 above)
Comment (tjoelsson) [reply defaults/main.yml]: The default value should be no

File: tasks/ssl_certificate.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+- name: create SSL certificate file
+  copy:
+    content: "{{ ssl_certs_crt }}"
+    dest: "{{ sp_https_certificate }}"
+    owner: root
+    group: root
+    mode: 0640
+
+- name: create symlink so php saml library can find the key file
```
Comment (tjoelsson) [inline tasks/ssl_certificate.yml]: This symlink is not needed. It's only for SAML implementation.

File: tasks/ssl_certificate.yml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+- name: create SSL key file
+  copy:
+    content: "{{ ssl_cert_priv_key }}"
+    dest: "{{ laravel_https_key }}"
+    owner: root
+    group: root
+    mode: 0640
+
+- name: create SSL certificate file
```
Comment (tjoelsson) [inline tasks/ssl_certificate.yml]: .sp file is also just for SAML. Can remove it.

File: tasks/ssl_certificate.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    group: root
+    mode: 0600
+
+- name: create SSL key file
+  copy:
+    content: "{{ ssl_cert_priv_key }}"
+    dest: "{{ laravel_https_key }}"
+    owner: root
+    group: root
+    mode: 0640
```
Comment (tjoelsson) [inline tasks/ssl_certificate.yml]: Change to 0600

**Review by tjoelsson — CHANGES_REQUESTED** [review]

File: defaults/main.yml [hunk 9]
```diff
@@ -5,6 +5,6 @@ deploy_group: '{{ ansible_ssh_user }}'
 laravel_rotate_nginx_log: true
 laravel_rotate_nginx_log_retention: 30
 
-laravel_https_enabled: no
+nginx_use_ssl: true
```
Comment (tjoelsson) [inline defaults/main.yml]: You can keep the name `laravel_https_enabled`. We want to have laravel_ as namespace for variables.

File: defaults/main.yml [hunk 10]
```diff
@@ -4,3 +4,7 @@ deploy_group: '{{ ansible_ssh_user }}'
 
 laravel_rotate_nginx_log: true
 laravel_rotate_nginx_log_retention: 30
+
+laravel_https_enabled: true
```
Comment (tjoelsson) [inline defaults/main.yml]: default to false

File: templates/admin.juwai.com.crt.j2 [hunk 11]
```diff
@@ -0,0 +1 @@
+{{ ssl_certs_crt }}{{ digicert_crt }}
```
Comment (tjoelsson) [inline templates/admin.juwai.com.crt.j2]: Use a generic name for this file, like `ssl.crt.j2`

File: tasks/ssl_certificate.yml [hunk 12]
```diff
@@ -0,0 +1,18 @@
+---
+- include_vars: "group_vars/env-production/vault.yml"
+
+- name: create SSL + digicert certificate file
+  template:
+    src: "{{ cert_source }}"
```
Comment (tjoelsson) [inline tasks/ssl_certificate.yml]: This should be the template file

File: tasks/main.yml [hunk 13]
```diff
@@ -88,6 +88,7 @@
     group: '{{ deploy_group }}'
   when: env == 'production' or env == 'staging'
 
+- include: ssl_certificate.yml
```
Comment (tjoelsson) [inline tasks/main.yml:91]: when laravel_https_enabled

**Review by tjoelsson — CHANGES_REQUESTED** [review]
Add all new variables to readme

**Comment by piadelosreyes** [comment]
Hi @tjoelsson, this is now ready for your review again. thanks

File: tasks/main.yml [hunk 14]
```diff
@@ -89,7 +89,7 @@
   when: env == 'production' or env == 'staging'
 
 - include: ssl_certificate.yml
-  when: laravel_https_enabled
+  when: laravel_https_enabled and (env == 'production' or env == 'staging')
```
Comment (tjoelsson) [inline tasks/main.yml]: No need to check environment here. The variable can be set for different environments in playbook inventory files.

File: templates/ssl.crt.j2 [hunk 15]
```diff
@@ -0,0 +1 @@
+{{ ssl_certs_crt }}{{ digicert_crt }}
```
Comment (tjoelsson) [inline templates/ssl.crt.j2]: There are two more variables here. Need to document. Better use the same naming conventions for them as well.

These variables contain the actual certificate content. Maybe the path variables should end with _path.

**Review by tjoelsson — CHANGES_REQUESTED** [review]

**Comment by piadelosreyes** [comment]
@tjoelsson please review again. thanks.

File: README.md [hunk 16]
```diff
@@ -77,6 +77,32 @@ Custom headers allowed by CORS, Access-Control-Allow-Headers.
...
+
+### laravel_https_key_path
+
+SSL certificate private key path
+
+### laravel_https_cert_data
+
+SSL certificate
+
+### laravel_https_digicert_data
```
Comment (tjoelsson) [inline README.md]: Digicert is a company providing SSL certificates. Let's make this more generic and call it `laravel_https_chain_data`. Chain as in the certificate chain, which can include intermediate and root certificates.

Comment (piadelosreyes) [reply README.md]: renamed in 05369c4

File: tasks/ssl_certificate.yml [hunk 17]
```diff
@@ -0,0 +1,18 @@
+---
+- include_vars: "group_vars/env-production/vault.yml"
+
+- name: create SSL + digicert certificate file
```
Comment (tjoelsson) [inline tasks/ssl_certificate.yml]: Just update this description and I think it's ok

**Review by tjoelsson — APPROVED** [review]

**Review by mark-juwai — APPROVED** [review]

**Pull request merged**